### DISCUSSION TITLE
Endpoints question.

### DISCUSSION THREAD

**Discussion by a-kbv** [Terrance/SkPy#233] [Q&A] [answered]
Can you provide some guidance on how to set presence and subscribe for endpoints in Skype? I am currently facing issues where I'm unable to set presence and subscribe for endpoints after retrieving data from a file. When logging in, I can successfully set presence and use the endpoint for message polling. However, when I try to do the same after retrieving data from a file, I receive 400 or 404 error codes. Can you also explain how setting presence and subscribing for endpoints work in Skype?
Here is some of my code:

# Function: configureEndpoint

This function is used to configure the endpoint by allowing presence and subscribing to the endpoint.

## Parameters
- `$session`: A Session object representing the current user's session.

## Code
```php
public function configureEndpoint(Session $session): void
{
    $this->allowPresence($session);
    $this->subscribeEndpoint($session);
    $this->subscribePresence($session);
}
```

# Function: allowPresence

This function is used to allow the presence of the endpoint.

## Parameters
- `$session`: A Session object representing the current user's session.

## Code
```php
public function allowPresence(Session $session): void
{
    $url = sprintf(
        '%s/users/ME/endpoints/%s/presenceDocs/messagingService',
        $session->getRegistrationToken()->getMessengerUrl(),
        $session->getEndpoint()->getId()
    );

    $request = null;
    try {
        $request = $this->request('PUT', $url, [
            'authorization_session' => $session,
            'headers' => [
                'Content-Type' => 'application/json',
                'Accept' => 'application/json',
            ],
            'body' => json_encode([
                'id' => 'messagingService',
                'type' => 'EndpointPresenceDoc',
                'selfLink' => 'uri',
                'privateInfo' => [
                    'epname' => "phpSkype",
                ],
                'publicInfo' => [
                    'capabilities' => '',
                    'type' => 1,
                    'skypeNameVersion' => 'skype.com',
                    'nodeInfo' => 'xx',
                    'version' => '908/1.30.0.128',
                ],
            ]),
        ]);
    } catch (ClientException $e) {
        $response = $e->getResponse();
        $response->getStatusCode();
    }

    if ($request->getStatusCode() == 200) {
        $result = json_decode($request->getContent(), true);
        $session->getEndpoint()->setPresence(true);
        $session->getEndpoint()->setName($result['privateInfo']['epname']);
        $session->getEndpoint()->setPresenceInfo($result);
    }
}
```

# Function: subscribeEndpoint

This function is used to subscribe to the endpoint.

## Parameters
- `$session`: A Session object representing the current user's session.

## Code
```php
public function subscribeEndpoint(Session $session): void
{
    $url = sprintf(
        '%s/users/ME/endpoints/%s/subscriptions',
        $session->getRegistrationToken()->getMessengerUrl(),
        $session->getEndpoint()->getId()
    );

    $resources = [
        '/v1/threads/ALL',
        '/v1/users/ME/contacts/ALL',
        '/v1/users/ME/conversations/ALL/messages',
        '/v1/users/ME/conversations/ALL/properties'
    ];

    $requestData = [
        'interestedResources' => $resources,
        'template' => 'raw',
        'channelType' => 'httpLongPoll',
        'conversationType' => 2047,
    ];

    $response = $this->request('POST', $url, [
        'authorization_session' => $session,
        'headers' => [
            'Content-Type' => 'application/json',
            'Accept' => 'application/json',
        ],
        'body' => json_encode($requestData),
    ]);

    if ($response->getStatusCode() == 201) {
        $session->getEndpoint()->setSubscribed(true);
        $session->getEndpoint()->setSubscribedUrl($response->getHeaders()['location'][0]);
    }
}
```

# Function: subscribePresence

This function is used to subscribe to presence.

## Parameters
- `$session`: A Session object representing the current user's session.

## Code
```php
public function subscribePresence(Session $session): void
{
    $url = sprintf(
        '%s/users/ME/endpoints/%s/subscriptions/0',
        $session->getRegistrationToken()->getMessengerUrl(),
        $session->getEndpoint()->getId()
    );

    $resources = [
        '/v1/threads/ALL',
        '/v1/users/ME/contacts/ALL',
        '/v1/users/ME/conversations/ALL/messages',
        '/v1/users/ME/conversations/ALL/properties'
    ];

    $contacts = $this->loadAllContacts($session);
    foreach ($contacts as $contact) {
        $resources[] = '/v1/users/ME/contacts/'. $contact->getPersonId();
    }

    try {
        $response = $this->request('PUT', $url, [
            'authorization_session' => $session,
            'query' =>
            [
                'name' => 'interestedResources',
            ],
            'json' =>[
                'interestedResources' => $resources,
            ]
        ]);
    } catch (ClientException $e) {
        $response = $e->getResponse();
        $response->getStatusCode();
    }
}
```

**Comment by Terrance** [comment]
From a cursory look, this looks roughly the same as what SkPy's [`SkypeEndpoint`](https://skpy.t.allofti.me/reference/internal.html#skpy.conn.SkypeEndpoint) does (which I assume is what this is based on), which itself is a best guess of what Skype for Web was doing around 2020.

> However, when I try to do the same after retrieving data from a file, I receive 400 or 404 error codes.

You're going to need to elaborate on that -- what file, where, how?

At a guess, that's taking a while and the endpoint has expired by the time you try to poll with it again.  See note on `SkypeEndpoint`:

https://github.com/Terrance/SkPy/blob/8f4d493ea5e89263e7b4f2bee33f0885c6cac994/skpy/conn.py#L1038

> Can you also explain how setting presence and subscribing for endpoints work in Skype?

(Again, this is all guesswork from Skype for Web.)  You get a default endpoint called `SELF`, and can create additional ones that come back with a UUID identifier.  Each of these can subscribe to any number of resources, which tells Skype which events you want to receive when you start listening for them via long-poll.  Your own availability is set by writing the `status` field of the endpoint (see [`Skype.setPresence`](https://skpy.t.allofti.me/reference/api.html#skpy.main.Skype.setPresence)).

**Reply by a-kbv** [reply]
Thank you for your fast answer and sorry for my late reply. The file I mentioned is a simple encrypted cache like "authentication data" stored in a file to prevent making many login attempts and too many requests for rarely changed data. Here is an example structure of the Model:

# Session Model

- **user**: Akbv\PhpSkype\Models\Account
  - username: "email@email"
  - password: "securepw"
- **skypeToken**: Akbv\PhpSkype\Models\SkypeToken
  - skypeToken: "eyJh..."
  - **expiry**: DateTime
    - date: "0001-01-01 01:01:01"
    - timezone: UTC
- **registrationToken**: Akbv\PhpSkype\Models\RegistrationToken
  - registrationToken: "U2lnb..."
  - **expiry**: DateTime
    - date: "0001-01-01 01:01:01"
    - timezone: "UTC"
  - messengerUrl: "https://azwcus1-client-s.gateway.messenger.live.com/v1"
- **securityToken**: Akbv\PhpSkype\Models\SecurityToken
  - .....
- **endpoint**: Akbv\PhpSkype\Models\Endpoint
  - id: "{64527ba5-....-9d40-72b4a41a3600}"
  - name: "{64527ba5-....-9d40-72b4a41a3600}"
  - presence: "true"
  - subscribed: "true"
  - subscribedPresence: "false"
  - presenceHost: "https://azwcus1-client-s.gateway.messenger.live.com/v1/users/8:live:myId/endpoints/{64527ba5-....-9d40-72b4a41a3600}/presenceDocs/messagingService"
- groups: Akbv\PhpSkype\Models\Group[]
  - ....
  
 After fresh login and when I execute the method `configureEndpoint` (it's written in my last question) I am able to execute my method to `getEvents` and everything works as expected, e.g I receive user presence info, new messages and etc. 
 ```PHP
    public function getEvents(Session $session): array
    {
        if ($session->getEndpoint()->getSubscribed() == false) {
            $this->subscribeEndpoint($session);
        }

        $url = sprintf(
            '%s/users/ME/endpoints/%s/subscriptions/0/poll',
            $session->getRegistrationToken()->getMessengerUrl(),
            $session->getEndpoint()->getId()
        );

        try{
            $response = $this->request('POST', $url, [
                'headers' => [
                    'Content-Type' => 'application/json',
                    'Accept' => 'application/json',
                ],
                'authorization_session' => $session,
            ]);
        } catch (ClientException $e) {
            $response = $e->getResponse();
            $response->getStatusCode();
        }

        $eventMessages = [];
        if (200 === $response->getStatusCode()) {
            $responseData = $response->toArray();
            foreach ($responseData['eventMessages'] as $eventMessage) {
                $eventType = isset($eventMessage['resourceType']) ? $eventMessage['resourceType'] : null;
                $eventMessages[] = [
                    'type' => $eventType,
                    'message' => $eventMessage,
                ];
            }
        }
        return $eventMessages;
    }
```
But when I come back after 5 mins, and run my code it extracts the data from the "authentication details" file and I try to execute `getEvents` with my endpoint an error is thrown with 404, and an error message appears to be in headers ['statustext'] : "You must create an endpoint before performing this operation.".
Does that mean this endpoint is "dead" because I have not made requests for 5 minutes to the endpoint? And now I have to create new one logging in again. Or do I have to somehow subscribe to it again?

As for the endpoint with `SELF`, it is returned in headers of method `subscribeEndpoint` "https://azwcus1-client-s.gateway.messenger.live.com/v1/users/ME/endpoints/SELF/subscriptions/0" 
If that is the host for creating a new endpoint, what parameters do I need to send in order to create a new endpoint after the old one is expired?

I am really struggling to understand how the flow with endpoints works, sorry if I'm asking stupid questions.
And yes I'm using SkPy as an example of workflow. As for now, it's a private project as I need to create Skype - Mattermost bridge. If it's successful I will open my library for public use and you and SkPy will be mentioned with many thanks and gratitude.

**Reply by Terrance** [reply, marked as answer]
> I come back after 5 mins

As noted above, your endpoint will have expired by then if you're not constantly polling it or calling the keep-alive API.  Just make a new endpoint and subscribe it again.

(It might be worth watching what Skype for Web does if you disconnect your computer for a few minutes.  I assume mobile clients handle this differently so that they're not constantly churning on endpoints, but they may have completely separate APIs.)

**Reply by a-kbv** [reply]
Thank you.
